### ISSUE TITLE
Deleting a selection can ruin proper nesting

### ISSUE THREAD

**Issue by xpe** [odyssomay/paredit#6]
```
(defn f1 [coll f x] (conj (map f coll) x))
```

If you select `x]` and press delete, you get:

```
(defn f1 [coll f (conj (map f coll) x))
```

The desired behavior is only to delete the `x`.

Note: If you selected something that includes an entire balanced expression, it would be ok to delete the balanced part of it, including parens.

I just checked this in Emacs ParEdit; it seems to allow an imbalance to occur. I find it more useful to keep everything balanced. Would others like to weigh in?

**Comment by odyssomay** [comment]
This was actually intentional.

The idea is that this way, it's fairly easy to fix unbalanced parens. That said, I'm definitely open to always keeping the balance. Maybe as an option, or as the default. Probably the latter.

**Comment by xpe** [comment]
Yes, but the main point of having ParEdit mode activated is to keep everything balanced. Normal delete keystrokes are balance-preserving. It seems consistent to say that selection-based deletes should be balance-preserving as well.

**Comment by odyssomay** [comment]
You're right, but I feel that too much "handholding" can lead to a lot of frustration. Normally, I know if the expression I want to remove has matched brackets, and if not, it was probably intentionally so. The plan right now is to keep things balanced though.

**Comment by odyssomay** [comment]
I added an option _strict_mode_ that tries to ensure as far as possible that parens are matched. Along with the implementation of _strict deleting_. There are some issues though, as can be seen in the tests:
Given: `(fo|o (bar|) baz)` The result is: `(fo|() baz)`

**Comment by xpe** [comment]
Great! `(fo|o (bar|) baz)` -> `(fo|() baz)` seems reasonable to me.

**Comment by odyssomay** [comment]
I actually changed it in the latest commit to result in `(fo|(bar) baz)`.

Consider:
`(fo|o {:a "baz" :b "ba|r"})`
This would result in
`(fo|{"r"})`
Which feels really odd, with the half-removed expressions like that. So the new behaviour only removes complete expressions in the selection. That is, the result would be:
`(fo|{:a "baz" :b "bar"})`
Because the map is not completely in the selection, none of it will be removed. I'm not sure if this is the best behaviour though, what do you think?

**Comment by xpe** [comment]
Either sound reasonable to me. Perhaps the latter is slightly less surprising. (The latter way could be summarized as "deletes the largest selected subexpression".)

Once more people are using this, they will weigh in too.

**Comment by odyssomay** [comment]
I agree. I'm reopening until we get more input as well as experience on how it feels to actually use in the real world.
